internal/builtin/data_yaml.go

```
@@ -0,0 +1,197 @@
+package builtin
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/zclconf/go-cty/cty"
+	"gopkg.in/yaml.v3"
+
+	"github.com/blackstork-io/fabric/pkg/diagnostics"
+	"github.com/blackstork-io/fabric/plugin"
+	"github.com/blackstork-io/fabric/plugin/dataspec"
+	"github.com/blackstork-io/fabric/plugin/plugindata"
+)
+
+type yamlData struct {
+	data plugindata.Data
+}
+
+func makeYAMLDataSource() *plugin.DataSource {
+	return &plugin.DataSource{
+		DataFunc: fetchYAMLData,
+		Args: &dataspec.RootSpec{
+			Attrs: []*dataspec.AttrSpec{
+				{
+					Name:       "glob",
+					Type:       cty.String,
+					ExampleVal: cty.StringVal("path/to/file*.yaml"),
+					Doc:        `A glob pattern to select YAML files to read`,
+				},
+				{
+					Name:       "path",
+					Type:       cty.String,
+					ExampleVal: cty.StringVal("path/to/file.yaml"),
+					Doc:        `A file path to a YAML file to read`,
+				},
+			},
+		},
+		Doc: `
+		Loads YAML files with the names that match provided ` + "`glob`" + ` pattern or a single file from provided ` + "`path`" + `value.
+
+		Either ` + "`glob`" + ` or ` + "`path`" + ` argument must be set.
+
+		When ` + "`path`" + ` argument is specified, the data source returns only the content of a file.
+		When ` + "`glob`" + ` argument is specified, the data source returns a list of dicts that contain the content of a file and file's metadata. For example:
+
+		` + "```yaml" + `
+		[
+		  {
+			"file_path": "path/file-a.yaml",
+			"file_name": "file-a.yaml",
+			"content": {
+			  "foo": "bar"
+			}
+		  },
+		  {
+			"file_path": "path/file-b.yaml",
+			"file_name": "file-b.yaml",
+			"content": [
+			  {"x": "y"}
+			]
+		  }
+		]
+		` + "```",
+	}
+}
+
+func fetchYAMLData(ctx context.Context, params *plugin.RetrieveDataParams) (plugindata.Data, diagnostics.Diag) {
+	glob := params.Args.GetAttrVal("glob")
+	path := params.Args.GetAttrVal("path")
```

Maybe we should throw an error when both attributes are set?